File: 1465-maximum-product-of-splitted-binary-tree/maximum-product-of-splitted-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxProduct = function(root) {
    const MAX = 50_001;
    const MOD = 1_000_000_007n;
    const memo = new Map()
    const dfs = (node) => {
        if (!node) return 0;
        let sum = node.val;
        sum += dfs(node.left);
        sum += dfs(node.right);
        memo.set(node, sum);
        return sum;
    }
    const total = dfs(root, 1);
    let maxProduct = 0n;

    for (const [key, value] of memo) {
        const now = BigInt(value) * BigInt(total - value);
        if (maxProduct < now) maxProduct = now;
    }

    return Number(maxProduct % MOD)
};

File: 1465-maximum-product-of-splitted-binary-tree/maximum-product-of-splitted-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxProduct(self, root: Optional[TreeNode]) -> int:
        MAX = 50_001
        MOD = 1_000_000_007
        memo = defaultdict(int)

        def dfs(node, index):
            if not node: return 0
            _sum = node.val
            _sum += dfs(node.left, index * 2)
            _sum += dfs(node.right, index * 2 + 1)
            memo[index] = _sum
            return _sum

        total = dfs(root, 1)
        maxProduct = 0
        for i, value in memo.items():
            maxProduct = max(maxProduct, (value * (total - value)))

        return maxProduct % MOD